### PR TITLE
RavenDB-19558 Introducing OrQueries optimization for Corax.

### PR DESCRIPTION
**Pull request by maciejaszyk**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19558

### Additional description

- Translate `(Field = x) or (Field = y) ...  or (Field = Zn)` into `(Field in (x, y, ..., Zn)`

### Type of change

- Optimization
- New feature

### How risky is the change?

- Moderate

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

**Review by ayende — APPROVED** [review]

**Comment by redknightlois** [comment]
LGTM

File: src/Corax/Queries/MemoizationMatch.cs [hunk 1]
```diff
@@ -57,14 +60,15 @@ public void Score(Span<long> matches, Span<float> scores)
...
 
         public QueryInspectionNode Inspect()
         {
             return new QueryInspectionNode($"{nameof(MemoizationMatch)} [Memoization]",
                 children: new List<QueryInspectionNode> { _inner.Inspect() },
                 parameters: new Dictionary<string, string>()
                 {
-                    { "BufferSize", _inner.Buffer.Length.ToString() },
+                   // { "BufferSize", _inner.Buffer.Length.ToString() },
                     { nameof(IsBoosting), IsBoosting.ToString() },
```
Comment (redknightlois) [inline src/Corax/Queries/MemoizationMatch.cs]: Remove this

File: src/Raven.Server/Documents/Indexes/Persistence/Corax/QueryOptimizer/CoraxAndQueries.cs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+            Reduce:
+            reduced++;
+
+        }
+
+        stack = stack.Slice(reduced);
+        if (stack.Length == 0)
+            goto Return;
+
+        // Ascending term amount stack. It not always would work for us. Thats is telling us terms inside field are clustered in centrains points
```
Comment (redknightlois) [inline src/Raven.Server/Documents/Indexes/Persistence/Corax/QueryOptimizer/CoraxAndQueries.cs]: typo?

File: src/Corax/Queries/MemoizationMatch.cs (hunk 1 above)
Comment (maciejaszyk) [reply src/Corax/Queries/MemoizationMatch.cs]: done

**Comment by arekpalinski** [comment]
@maciejaszyk can you please check failing tests?

**Pull request merged**